File: common/src/main/java/com/luminaryn/common/Images.java
package com.luminaryn.common;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Base64;

import androidx.annotation.NonNull;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Objects;

public class Images {

    private static void saveImage(@NonNull final Context context, @NonNull final Bitmap bitmap,
                           @NonNull final Bitmap.CompressFormat format, @NonNull final String mimeType,
                           @NonNull final int quality, @NonNull final String filename) throws IOException {
        OutputStream fos;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            ContentResolver resolver = context.getContentResolver();
            ContentValues contentValues = new ContentValues();
            contentValues.put(MediaStore.MediaColumns.DISPLAY_NAME, filename);
            contentValues.put(MediaStore.MediaColumns.MIME_TYPE, mimeType);
            contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_PICTURES);
            Uri imageUri = resolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, contentValues);
            fos = resolver.openOutputStream(Objects.requireNonNull(imageUri));
        } else {
            String imagesDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString();
            File image = new File(imagesDir, filename);
            fos = new FileOutputStream(image);
        }
        bitmap.compress(format, quality, fos);
        Objects.requireNonNull(fos).close();
    }

    public static void saveImage(Context context, Bitmap bitmap, @NonNull int quality, @NonNull String filename) throws IOException {
        final Bitmap.CompressFormat format = Bitmap.CompressFormat.JPEG;
        final String mimeType = "image/jpeg";
        saveImage(context, bitmap, format, mimeType, quality, filename);
    }

    public static void saveImage(Context context, Bitmap bitmap, @NonNull String filename) throws IOException {
        saveImage(context, bitmap, 100, filename);
    }

    public static String base64Encode (Bitmap bitmap, @NonNull Bitmap.CompressFormat format, @NonNull int quality) throws IOException {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        bitmap.compress(format, quality, bos);
        String base64 = Base64.encodeToString(bos.toByteArray(), Base64.DEFAULT);
        bos.close();
        return base64;
    }

    public static String base64Encode (Bitmap bitmap, @NonNull int quality) throws IOException {
        return base64Encode(bitmap, Bitmap.CompressFormat.JPEG, quality);
    }

    public static String base64Encode (Bitmap bitmap) throws IOException {
        return base64Encode(bitmap, 100);
    }

    public static byte[] toJPEG (Bitmap bitmap, @NonNull int quality) {
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, stream);
        return stream.toByteArray();
    }

    public static byte[] toJPEG (Bitmap bitmap) {
        return toJPEG(bitmap, 100);
    }

}

File: common/src/main/java/com/luminaryn/common/Images.kt
package com.luminaryn.common

import android.content.ContentValues
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.util.Base64
import java.io.*
import java.lang.Exception
import java.util.*

object Images {
    @JvmOverloads
    @JvmStatic
    @Throws(IOException::class)
    fun saveImage(bitmap: Bitmap,
                  file: File,
                  format: CompressFormat = CompressFormat.JPEG,
                  quality: Int = 100) {
        saveImage(bitmap, FileOutputStream(file), format, quality)
    }

    @JvmOverloads
    @JvmStatic
    fun saveImage(bitmap: Bitmap,
                  output: OutputStream,
                  format: CompressFormat = CompressFormat.JPEG,
                  quality: Int = 100) {
        bitmap.compress(format, quality, output)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun saveImage(context: Context, bitmap: Bitmap,
                  format: CompressFormat, mimeType: String,
                  quality: Int, filename: String) {
        val fos: OutputStream?
        fos = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val resolver = context.contentResolver
            val contentValues = ContentValues()
            contentValues.put(MediaStore.MediaColumns.DISPLAY_NAME, filename)
            contentValues.put(MediaStore.MediaColumns.MIME_TYPE, mimeType)
            contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_PICTURES)
            val imageUri = resolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, contentValues)
            imageUri?.let { resolver.openOutputStream(it) }
        } else {
            val imagesDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString()
            val image = File(imagesDir, filename)
            FileOutputStream(image)
        }
        bitmap.compress(format, quality, fos)
        Objects.requireNonNull(fos)?.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    fun saveImage(context: Context, bitmap: Bitmap, quality: Int, filename: String) {
        val format = CompressFormat.JPEG
        val mimeType = "image/jpeg"
        saveImage(context, bitmap, format, mimeType, quality, filename)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun saveImage(context: Context, bitmap: Bitmap, filename: String) {
        saveImage(context, bitmap, 100, filename)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun base64Encode(bitmap: Bitmap, format: CompressFormat, quality: Int): String {
        val bos = ByteArrayOutputStream()
        bitmap.compress(format, quality, bos)
        val base64 = Base64.encodeToString(bos.toByteArray(), Base64.DEFAULT)
        bos.close()
        return base64
    }

    @JvmStatic
    @JvmOverloads
    @Throws(IOException::class)
    fun base64Encode(bitmap: Bitmap, quality: Int = 100): String {
        return base64Encode(bitmap, CompressFormat.JPEG, quality)
    }

    @JvmStatic
    fun toJPEGStream(bitmap: Bitmap, quality: Int): ByteArrayOutputStream {
        val stream = ByteArrayOutputStream()
        bitmap.compress(CompressFormat.JPEG, quality, stream)
        return stream
    }

    @JvmStatic
    @JvmOverloads
    fun toJPEG(bitmap: Bitmap, quality: Int = 100): ByteArray {
        return toJPEGStream(bitmap, quality).toByteArray()
    }

    fun fromBase64(string: String): Bitmap? {
        val bytes = Base64.decode(string, Base64.DEFAULT)
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }
}